extensions/amp-vi-player/0.1/amp-vi-player.js

```diff
@@ -0,0 +1,225 @@
+    const consentPolicyStatePromise = getConsentPolicyState(
+      this.element,
+      consentPolicyId
+    );
+    const consentValuePromise = getConsentPolicyInfo(
+      this.element,
+      consentPolicyId
+    );
+
+    return Promise.all([
+      consentPolicySharedDataPromise,
+      consentPolicyStatePromise,
+      consentValuePromise,
+    ]);
+  }
+
+  /**
+   * Listen to iframe messages and handle events
+   *
+   * @param {!Event} event
+   * @private
+   */
+  handleIframeMessages_(event) {
+    if (!this.iframe_ || event.source !== this.iframe_.contentWindow) {
+      // Ignore messages from other iframes.
+      return;
+    }
+    const data = getData(event);
+    if (
+      data &&
+      data['message'] === 'vi-stories-unit-event' &&
+      data['action'] === 'close'
+    ) {
+      this.destroyPlayerIframe_();
+      this.attemptCollapse();
+    }
+  }
+
+  /**
+   * Remove the player iframe
+   * @private
+   */
+  destroyPlayerIframe_() {
+    this.win.removeEventListener('message', this.boundHandleIframeMessages_);
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+  }
+
+  /**
+   * @param {boolean=} onLayout
+   * @override
+   */
+  preconnectCallback(onLayout) {
+    // Serves the player assets
+    Services.preconnectFor(this.win).url(
+      this.getAmpDoc(),
+      this.viDomain_,
+      onLayout
+    );
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  buildCallback() {
+    const {element, win} = this;
+
+    setIsMediaComponent(element);
+
+    for (const key in element.dataset) {
+      this.params_[key] = element.dataset[key];
+    }
+
+    // Publisher id is mandatory
+    userAssert(
+      this.params_['publisherId'],
+      'The data-publisher-id attribute is required for <amp-vi-player> %s',
+      element
+    );
+
+    // Channel id is mandatory
+    userAssert(
+      this.params_['channelId'],
+      'The data-channel-id attribute is required for <amp-vi-player> %s',
+      element
+    );
+
+    win.addEventListener('message', this.boundHandleIframeMessages_);
+  }
+
+  /** @override */
+  layoutCallback() {
+    const timeout = new Promise(function (resolve) {
+      setTimeout(() => resolve([]), 3000);
+    });
+
+    return Promise.race([timeout, this.getConsentData_()]).then((consents) => {
+      const {element, win} = this;
+      const src = addParamsToUrl(
+        this.viDomain_ + '/tagLoaderAmp.html',
+        this.params_
+      );
+      const attributes = getContextMetadata(
+        win,
+        element,
+        generateSentinel(win),
+        dict({
+          'viTitle': document.title,
+          'viPageLanguage': win.document.documentElement.lang,
+        })
+      );
+      Object.assign(attributes['_context'], {
+        'consentSharedData': consents[0],
+        'initialConsentState': consents[1],
+        'initialConsentValue': consents[2],
+      });
```

Comment:
Can we add a test to ensure that the consent information shows up in the iframe attributes as expected?

Reply:
Done. Please have a look now.